cmd/utils.go

```diff
@@ -67,28 +67,37 @@ func request2BucketObjectName(r *http.Request) (bucketName, objectName string) {
 	if err != nil {
 		logger.CriticalIf(context.Background(), err)
 	}
-	return urlPath2BucketObjectName(path)
-}
-
-// Convert url path into bucket and object name.
-func urlPath2BucketObjectName(path string) (bucketName, objectName string) {
-	if path == "" || path == SlashSeparator {
-		return "", ""
-	}
 
-	// Trim any preceding slash separator.
-	urlPath := strings.TrimPrefix(path, SlashSeparator)
-
-	// Split urlpath using slash separator into a given number of
-	// expected tokens.
-	tokens := strings.SplitN(urlPath, SlashSeparator, 2)
-	bucketName = tokens[0]
-	if len(tokens) == 2 {
-		objectName = tokens[1]
+	return path2BucketObject(path)
+}
+
+func path2BucketObject(s string) (bucket, prefix string) {
+	tokens := make([]string, 2)
+	return path2BucketObjectWithTokens("", s, tokens)
+}
+
+// Similar to https://golang.org/pkg/strings/#SplitN
+// but doesn't trucate the return list, keeps the
+// size of the array 'n'
+func path2BucketObjectWithTokens(b, s string, tokens []string) (bucket, prefix string) {
```

Comment:
Simpler:

```Go
// path2BucketObjectWithTokens returns bucket and prefix, if any, of a path 's'.
// b is trimmed from the front of the path.
// (not sure where the tokens fit into this)
func path2BucketObjectWithTokens(b, s string) (bucket, prefix string) {
	s = strings.TrimPrefix(s, b)
	s = strings.TrimPrefix(s, SlashSeparator)
	m := strings.Index(s, SlashSeparator)
	if m < 0 {
		return s, ""
	}
	return s[:m], s[m+len(SlashSeparator):]
}
```

Allocations are still only on the stack and the caller does not have to allocate the slice. The slice allocation is not included in the benchmark of the reference.

```
BenchmarkSlash/ref-12         	66847153	        18.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkSlash/simple-12      	100000000	        12.9 ns/op	       0 B/op	       0 allocs/op
```

Reply:
This is neat @klauspost